AI #1803: Contract Commitment dataset extension (#1924)
FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec: specification/datasets/contract_commitment/columns/contractcommitmentdiscountpercentage.md

### Tiered Incentives

For commitments with multiple tiers (e.g., 5% discount up to 1M, 10% above 1M), this column should represent the **active** or **base** discount percentage applicable to the current contract row.

## Requirements

ContractCommitmentDiscountPercentage adheres to the following requirements:

* ContractCommitmentDiscountPercentage MUST be present in a Contract Commitment [*FOCUS dataset*](#glossary:FOCUS-dataset).
* ContractCommitmentDiscountPercentage MUST be of type Decimal.
* ContractCommitmentDiscountPercentage MUST conform to [NumericFormat](#attributes.numericformat) requirements.
* ContractCommitmentDiscountPercentage MUST NOT be null if [ContractCommitmentBenefitCategory](#datasets.contractcommitment.contractcommitmentbenefitcategory) is "Discount".
* ContractCommitmentDiscountPercentage MUST be null if ContractCommitmentBenefitCategory is "Credit" or "Availability Guarantee" and no secondary discount is applied.
* ContractCommitmentDiscountPercentage MUST be a value between 0.0 and 1.0.

> * For contracts with multiple tiers (e.g., 5% discount up to 1M, 10% above 1M), ContractCommitmentDiscountPercentage MUST represent the discount percentage applicable to the current contract commitment.

Reviewer: In the case that we have a tiered discount then, I would not "learn" of the tier until I trigger it? I may not be understanding this entirely. This gets to the transactional nature of records in this dataset vs. it serving as more of a dataset. 

I don't know how to resolve this. I will accept this as an MVP but think it will need to be addressed more in future.

Author: Generally, I presume the tiers would be handled by representing each one as its own record.

Extending the example in the requirement: imagine the 10% tier exists in the dataset on Day 1 with a Status of `Pending`. The data generator doesn't change the 5% row into a 10% row; they simply flip the Status of the second row to `Active`. This keeps the dataset as a static 'source of truth' for the entire agreement while still allowing for automated logic.

Does that assuage your concern or no?

Author: Resolving, given Rob's approval of the PR.

### Calculating the Effective Percentage

In scenarios where a commitment has "stacked" or "nested" discounts, this field should reflect the total reduction from the list price.

Discount Percentage = 1 - (Contracted Unit Price / List Unit Price)

### Relationship to Benefit Category

This field provides the magnitude for the **Discount** category. While `ContractCommitmentCost` tracks the financial obligation (what you pay), `ContractCommitmentDiscountPercentage` tracks the benefit rate (what you save).

### Tiered Incentives

For commitments with multiple tiers (e.g., 5% discount up to 1M, 10% above 1M), this column should represent the **active** or **base** discount percentage applicable to the current contract row.

## Requirements

ContractCommitmentDiscountPercentage adheres to the following requirements:

* ContractCommitmentDiscountPercentage MUST be of type Decimal.
* ContractCommitmentDiscountPercentage MUST conform to [NumericFormat](#attributes.numericformat) requirements.
* ContractCommitmentDiscountPercentage MUST NOT be null if [ContractCommitmentBenefitCategory](#datasets.contractcommitment.contractcommitmentbenefitcategory) is "Discount".

> * ContractCommitmentDiscountPercentage MUST be null if ContractCommitmentBenefitCategory is "Credit" or "Availability Guarantee" and no secondary discount is applied.

Reviewer 2: fwiw having terms like "and no secondary discount is applied" make these rules Dynamic as its not defined what it means to have a secondary discount is applied. A way we could avoid this—maybe for a future release—is to have a way to flag that a secondary discount applies making rules like this to become 

```
* ContractCommitmentDiscountPercentage MUST be null if ContractCommitmentBenefitCategory is "Credit" or "Availability Guarantee", and ContractCommitmentSecondaryBenefit is false.
```

Author: I'm happy to simplify this and cut the language about secondary discount.  I think it still works in that form.  What do you think?

(I also see that it's still using the full term "Availability Guarantee", but in a more recent revision, I cut this down to "Availability".  Shrinking that to match here.)

```suggestion
* ContractCommitmentDiscountPercentage MUST be null if ContractCommitmentBenefitCategory is "Credit" or "Availability".
```

* ContractCommitmentDiscountPercentage MUST be a value between 0.0 and 1.0.

> * For contracts with multiple tiers (e.g., 5% discount up to 1M, 10% above 1M), ContractCommitmentDiscountPercentage MUST represent the discount percentage applicable to the current contract commitment.

Reviewer 2: " the discount percentage applicable to the current contract commitment." is a little ambiguous to me. does this mean you have multiple entires in the contract commitment table for each tier and the ContractCommitmentDiscountPercentage applies to the tier being described in the row. or does this mean the provider updates the ContractCommitmentDiscountPercentage based on which one is currently active (you made it to tier 2 so the ContractCommitmentDiscountPercentage updates in your Contract Commitment dataset?

Author: I intended for it to mean the former.  How might you suggest changing this to make that more clear?

Reviewer 2: Maybe something like

ContractCommitmentDiscountPercentage MUST reflect the discount percentage defined for the specific pricing tier represented by the Contract Commitment entry. For contracts with multiple pricing tiers (e.g., 5% discount up to 1M, 10% above 1M), each tier MUST be represented as a separate Contract Commitment entry.

Author: I like that -- but that's two normative statements.  Should we break them into two rows?

Reviewer 2: When ContractCommitmentDiscountPercentage applies to a contract with multiple pricing tiers (e.g., 5% discount up to 1M, 10% above 1M), ContractCommitmentDiscountPercentage MUST adhere to the following additional requirements:
	•	ContractCommitmentDiscountPercentage MUST reflect the discount percentage defined for the specific pricing tier represented by the Contract Commitment entry.
	•	ContractCommitmentDiscountPercentage MUST correspond to only one pricing tier per Contract Commitment entry.

Author: Thank you!  Here is the verbiage I went with:

* For contracts with multiple tiers (e.g., 5% discount up to 1M, 10% above 1M), ContractCommitmentDiscountPercentage MUST adhere to the following additional requirements:
  * ContractCommitmentDiscountPercentage MUST reflect the discount percentage defined for the specific pricing tier represented by the Contract Commitment row.
  * ContractCommitmentDiscountPercentage MUST correspond to only one pricing tier per Contract Commitment row.

Resolving.  Please reopen if that doesn't work for you.

> * ContractCommitmentDiscountPercentage SHOULD represent the net effective discount if multiple contractual layers are applicable (e.g., a negotiated discount on top of a standard commitment).

Reviewer 2: I think we should include a net effective discount item in the examples 

```
For example:
	•	Standard commitment discount: 20%
	•	Additional negotiated discount: 10% applied to the already-discounted price

The effective discount is not 30%.
It is:
1 - (0.8 × 0.9) = 28%

In this scenario, ContractCommitmentDiscountPercentage SHOULD be reported as 28%, representing the net effective discount.

```

Author: I added the example to the Implementation Guidance section.  What do you think?

Author: @mike-finopsorg Going to resolve this.  Feel free to reopen if you feel we need to review further.

# Contract Commitment Discount Percentage

> Contract Commitment Discount Percentage represents the effective percentage reduction applied to the list price for resources or services covered by a [*contract commitment*](#glossary:contract-commitment).

Reviewer 3: This should be tied to the specific resource or service, but is written like it applies to the entire contract - shouldn't this link to "ContractCommitmentEligibility" instead of "resources or service"?

Example:
- Customer could negotiate different discounts for EC2, S3, DynamoDB, different SavingsPlans rate for EC2 than RDS, S3 storage vs. network costs, etc.  It can get very detailed.

Author: A Contract can have several Contract Commitments.  This is why this entity has a `ContractId` and a `ContractCommitmentId`.  Each of the discounts in your example would be represented by separate rows in this dataset.

I am happy to discuss this further with you, either in this thread or via a sync.  But I think you may be missing the high-level scope of this dataset; I think you should read the dataset definition, which will either clarify your understanding (in which case, great), or it won't (in which case, we should consider augmenting that definition).  But making a change here in Contract Commitment Discount Percentage, I believe, is not warranted.

Will leave open until I hear back from you on that.

Reviewer 3: All good.

* Negotiated discount: 10% applied to the post-commitment discount cost

The effective discount is not 30%.
It is:
1 - (0.8 × 0.9) = 28%

In this scenario, ContractCommitmentDiscountPercentage should be reported as 28%.

### Relationship to Benefit Category

This field provides the magnitude for the **Discount** category. While `ContractCommitmentCost` tracks the financial obligation (what you pay), `ContractCommitmentDiscountPercentage` tracks the benefit rate (what you save).

### Tiered Incentives

For commitments with multiple tiers (e.g., 5% discount up to 1M, 10% above 1M), this column should represent the **active** or **base** discount percentage applicable to the current contract row.

## Requirements

ContractCommitmentDiscountPercentage adheres to the following requirements:

* ContractCommitmentDiscountPercentage MUST be of type Decimal.
* ContractCommitmentDiscountPercentage MUST conform to [NumericFormat](#attributes.numericformat) requirements.
* ContractCommitmentDiscountPercentage MUST NOT be null if [ContractCommitmentBenefitCategory](#datasets.contractcommitment.contractcommitmentbenefitcategory) is "Discount".

> * ContractCommitmentDiscountPercentage MUST be null if ContractCommitmentBenefitCategory is "Credit" or "Availability".

Reviewer 4: ```suggestion
* ContractCommitmentDiscountPercentage MUST be null if ContractCommitmentBenefitCategory is "Monetary Pool" or "Availability".
```

"Credit" doesn't appear in the ContractCommitmentBenefitCategory allowed values (Discount, Monetary Pool, Availability, Other). I'd recommend "Monetary Pool" here since the intent appears to be nulling out the discount percentage for non-discount benefit types.

Author: Yes, good call.  We changed Credit to Monetary Pool per TF-1 feedback, but I didn't circle back and scrub this reference.

> * ContractCommitmentDiscountPercentage MUST be a value between 0.0 and 1.0.

Author: ```suggestion
* ContractCommitmentDiscountPercentage MUST be a value between 0.0 and 1.0, inclusive.
```

* For contracts with multiple tiers (e.g., 5% discount up to 1M, 10% above 1M), ContractCommitmentDiscountPercentage MUST adhere to the following additional requirements:
  * ContractCommitmentDiscountPercentage MUST reflect the discount percentage defined for the specific pricing tier represented by the Contract Commitment row.
  * ContractCommitmentDiscountPercentage MUST correspond to only one pricing tier per Contract Commitment row.

> * ContractCommitmentDiscountPercentage SHOULD represent the net effective discount if multiple contractual layers are applicable (e.g., a negotiated discount on top of a standard commitment).

Reviewer 5: **Note:** This is probably **not in scope for this PR**, but I wish to flag it for **future consideration**.

It is currently unclear which cost metrics in the Cost and Usage dataset this discount applies to or should be reflected in.

Based on the rest of the spec and our previous discussions, these would be the expectations:

* Publicly available volume-based discounts MUST be reflected in all cost metrics (given that SKU Price ID reflects the tier)
* EDP and MACC discounts MUST NOT be reflected in `ListCost` / `ListUnitPrice`
* RI and Savings Plan discounts MUST NOT be reflected in `ListCost` / `ListUnitPrice` and `ContractedCost` / `ContractedUnitPrice`

### Question:

That said, should **this PR** explicitly require that ContractCommitmentDiscountPercentage  must be reflected (included) in `EffectiveCost` and `BilledCost` (even when `BilledCost` is 0, i.e., the charge is fully invoiced on a corresponding *covering* charge), the discount can still be considered applied)?

Author: I see your point.  I could go either way -- but, given the ongoing discussion of EffectiveCost and BilledCost per PR #1966, I think I'd rather leave it like this for now.  What do you think?

Author: Discussed with @ijurica.  No action for now.  More clarity potentially added in future release.

## Column ID

ContractCommitmentDiscountPercentage

## Display Name

Contract Commitment Discount Percentage

## Description

The effective percentage reduction applied to the list price of resources or services covered by a [*contract commitment*](#glossary:contract-commitment).

## Content Constraints

| Constraint      | Value          |
| :-------------- | :------------- |
| Dataset         | [Contract Commitment](#datasets.contractcommitment)  |
| Column type     | Dimension      |
| Feature level   | Mandatory      |
| Allows nulls    | True           |
| Data type       | Decimal        |
| Value format    | [Numeric Format](#attributes.numericformat) |

> | Number range    | 0.0 to 1.0     |

Author: ```suggestion
| Number range    | 0.0 to 1.0 (inclusive)     |
```
